classes/steps/actions/webservice_action_step.php

```diff
@@ -0,0 +1,262 @@
+
+        // Who.
+        $attributes = ['placeholder' => 'username'];
+        $mform->addElement('text', 'username', get_string('webserviceactionusername', 'tool_trigger'), $attributes);
+        $mform->setType('username', PARAM_ALPHANUMEXT);
+        $mform->addHelpButton('username', 'webserviceactionusername', 'tool_trigger');
+
+        // Params.
+        $attributes = ['cols' => '50', 'rows' => '5'];
+        $mform->addElement('textarea', 'params', get_string('webserviceactionparams', 'tool_trigger'), $attributes);
+        $mform->setType('params', PARAM_RAW_TRIMMED);
+        $mform->addHelpButton('params', 'webserviceactionparams', 'tool_trigger');
+    }
+
+    /**
+     * {@inheritDoc}
+     * @see \tool_trigger\steps\base\base_step::add_privacy_metadata()
+     */
+    public static function add_privacy_metadata($collection, $privacyfields) {
+        return $collection->add_external_location_link(
+            'webservice_action_step',
+            $privacyfields,
+            'step_action_webservice:privacy:desc'
+        );
+    }
+
+    /**
+     * Get a list of fields this step provides.
+     *
+     * @return array $stepfields The fields this step provides.
+     */
+    public static function get_fields() {
+        return self::$stepfields;
+    }
+
+    public function form_validation($data, $files) {
+        global $DB;
+
+        $errors = [];
+
+        // Check if the username links to a valid user, if set.
+        if (!empty($data['username'])) {
+            try {
+                $DB->get_record('user', ['username' => $data['username']], '*', MUST_EXIST);
+            } catch (\Throwable $e) {
+                $errors['username'] = $e->getMessage();
+            }
+        }
+
+        // Check if the provided function (name) is a valid callable function.
+        if (!empty($data['functionname'])) {
+            try {
+                $errorfield = 'functionname';
+                $function = \external_api::external_function_info($data['functionname']);
+
+                $errorfield = 'params';
+
+                // Fill template fields with a number.
+                $transformcallback = function() {
+                    return 0;
+                };
+
+                // Cannot use redner_datafields since we need to know of the
+                // datafields in advance. Will need to apply the change
+                // manually.
+                $params = preg_replace_callback(
+                    $this->datafieldregex,
+                    $transformcallback,
+                    $data['params']
+                );
+
+                // Check if this is valid JSON before doing the function's validate_parameters check.
+                $preparedparams = [];
+                if (!empty($params)) {
+                    $preparedparams = json_decode($params, true);
+                    if (is_null($preparedparams)) {
+                        throw new \Exception('Invalid Syntax');
```

Comment:
Note it's currently not a moodle lang string because this is the error message that bubbles up normally in > php 7.2.

Happy to add it if you think it's necessary.